src/main/kotlin/com/makarytskyi/rentcar/service/impl/OrderServiceImpl.kt

```diff
@@ -22,69 +27,83 @@ internal class OrderServiceImpl(
-
-    override fun patch(id: String, orderRequest: UpdateOrderRequest): OrderResponse {
-        val order = orderRepository.findById(id) ?: throw NotFoundException("Order with $id is not found")
-        val newFrom = orderRequest.from ?: order.from
-        val newTo = orderRequest.to ?: order.to
-        validateDates(newFrom!!, newTo!!)
-        validateCarAvailability(order.car?.id.toString(), newFrom, newTo)
-
-        return orderRepository.patch(id, UpdateOrderRequest.toPatch(orderRequest))
-            ?.let { OrderResponse.from(it, order.car?.price) }
-            ?: throw NotFoundException("Order with $id is not found")
-    }
-
-    private fun validateDates(from: Date, to: Date) {
-        require(to.after(from)) { "Start date must be before end date" }
-        require(from.after(Date())) { "Dates must be in future" }
+            .flatMap { Mono.just(it).zipWith(getCarPrice(it.carId.toString())) }
+            .map { OrderResponse.from(it.t1, it.t2) }
+
+    override fun patch(id: String, orderRequest: UpdateOrderRequest): Mono<OrderResponse> =
+        orderRepository.findFullById(id)
+            .switchIfEmpty { Mono.error(NotFoundException("Order with $id is not found")) }
+            .flatMap {
+                val from = orderRequest.from ?: it.from
+                val to = orderRequest.to ?: it.to
+                validateDates(from, to)
+                validateCarAvailability(it.car?.id.toString(), from, to)
+            }
+            .switchIfEmpty { orderRepository.patch(id, UpdateOrderRequest.toPatch(orderRequest)) }
+            .flatMap { order -> getCarPrice(order.carId.toString()).map { OrderResponse.from(order, it) } }
+
+    private fun validateDates(from: Date?, to: Date?) {
+        require(to?.after(from) == true) { "Start date must be before end date" }
+        require(from?.after(Date()) == true) { "Dates must be in future" }
     }
 
-    private fun validateUserExists(userId: String) =
-        require(userRepository.findById(userId) != null) { "User with id $userId is not found" }
-
-    private fun validateCarAvailability(carId: String?, from: Date, to: Date) {
-        require(carId != null) { "Car id should not be null" }
-        carRepository.findById(carId) ?: throw NotFoundException("Car with id $carId is not found")
-
-        val carOrders = orderRepository.findByCarId(carId)
-
-        require(carOrders.none { it.from?.before(to) == true && it.to?.after(from) == true }) {
-            "Order on these dates is already exist"
-        }
+    private fun validateUserExists(userId: String) = userRepository.findById(userId)
+        .switchIfEmpty { Mono.error(IllegalArgumentException("User with id $userId is not found")) }
+
+    private fun validateCarAvailability(carId: String?, from: Date?, to: Date?): Mono<MongoOrder> {
+        return Mono.justOrEmpty(carId)
+            .switchIfEmpty { Mono.error(IllegalArgumentException("Car id should not be null")) }
```

Comment:
there's nothing reactive here. IMO  such code can be written without wrapping ID to mono